File: search/array-based-search/interpolation-search/cpp/interpolation_search_test.cpp
#include <gtest/gtest.h>
#include <vector>

// Forward declaration
int interpolationSearch(const std::vector<int>& arr, int target);

// Test: Target found in uniform array
TEST(InterpolationSearchTest, TargetFoundInUniformArray) {
    std::vector<int> arr = {10, 20, 30, 40, 50, 60, 70, 80, 90};
    EXPECT_EQ(interpolationSearch(arr, 70), 6) << "Should find 70 at index 6";
}

// Test: Target at beginning
TEST(InterpolationSearchTest, TargetAtBeginning) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(interpolationSearch(arr, 1), 0) << "Should find first element";
}

// Test: Target at end
TEST(InterpolationSearchTest, TargetAtEnd) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(interpolationSearch(arr, 5), 4) << "Should find last element";
}

// Test: Not present
TEST(InterpolationSearchTest, TargetNotFound) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(interpolationSearch(arr, 6), -1) << "Should return -1 for missing element";
}

// Test: Empty array
TEST(InterpolationSearchTest, EmptyArray) {
    std::vector<int> arr = {};
    EXPECT_EQ(interpolationSearch(arr, 10), -1) << "Empty array should return -1";
}

// Test: Single element found
TEST(InterpolationSearchTest, SingleElementFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(interpolationSearch(arr, 5), 0) << "Single element found → index 0";
}

// Test: Single element not found
TEST(InterpolationSearchTest, SingleElementNotFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(interpolationSearch(arr, 10), -1) << "Single element not found → -1";
}

// Test: Duplicate values (all same)
TEST(InterpolationSearchTest, DuplicateValuesAllSame) {
    std::vector<int> arr = {5, 5, 5, 5};
    EXPECT_EQ(interpolationSearch(arr, 5), 0) << "All duplicates — match returns index 0";
    EXPECT_EQ(interpolationSearch(arr, 3), -1) << "All duplicates — no match returns -1";
}

// Test: Non-uniform data (worst-case-like)
TEST(InterpolationSearchTest, NonUniformDataStillFinds) {
    std::vector<int> arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 1000};
    EXPECT_EQ(interpolationSearch(arr, 9), 8) << "Should still find even in skewed data";
}

// Test: Division-by-zero protection
TEST(InterpolationSearchTest, DivisionByZeroProtection) {
    std::vector<int> arr = {7, 7, 7, 7};
    EXPECT_EQ(interpolationSearch(arr, 7), 0) << "Equal elements — handled safely";
    EXPECT_EQ(interpolationSearch(arr, 5), -1) << "Equal elements — not found returns -1";
}

// Test: Bounds safety (extrapolation fails gracefully)
TEST(InterpolationSearchTest, OutOfBoundsEstimationSafe) {
    std::vector<int> arr = {10, 20, 30};
    int result = interpolationSearch(arr, 25);
    EXPECT_TRUE(result == -1 || result == 2) << "Should not crash on bad estimate";
}

// Entry point for tests
int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: search/array-based-search/interpolation-search/java/InterpolationSearchTest.java
import org.junit.Test;
import static org.junit.Assert.*;

/**
 * Unit tests for the InterpolationSearch class.
 */
public class InterpolationSearchTest {

    @Test
    public void testTargetFoundInUniformArray() {
        assertEquals("Should find 70 at index 6", 6, InterpolationSearch.interpolationSearch(new int[]{10, 20, 30, 40, 50, 60, 70, 80, 90}, 70));
    }

    @Test
    public void testTargetAtBeginning() {
        assertEquals("Should find first element", 0, InterpolationSearch.interpolationSearch(new int[]{1, 2, 3, 4, 5}, 1));
    }

    @Test
    public void testTargetAtEnd() {
        assertEquals("Should find last element", 4, InterpolationSearch.interpolationSearch(new int[]{1, 2, 3, 4, 5}, 5));
    }

    @Test
    public void testTargetNotPresent() {
        assertEquals("Should return -1 for missing element", -1, InterpolationSearch.interpolationSearch(new int[]{1, 2, 3, 4, 5}, 6));
    }

    @Test
    public void testEmptyArray() {
        assertEquals("Empty array should return -1", -1, InterpolationSearch.interpolationSearch(new int[]{}, 10));
    }

    @Test
    public void testSingleElementFound() {
        assertEquals("Single element found → index 0", 0, InterpolationSearch.interpolationSearch(new int[]{5}, 5));
    }

    @Test
    public void testSingleElementNotFound() {
        assertEquals("Single element not found → -1", -1, InterpolationSearch.interpolationSearch(new int[]{5}, 10));
    }

    @Test
    public void testDuplicateValuesAllSame() {
        assertEquals("All duplicates — should return index 0 for match", 0, InterpolationSearch.interpolationSearch(new int[]{5, 5, 5, 5}, 5));
        assertEquals("All duplicates — should return -1 for no match", -1, InterpolationSearch.interpolationSearch(new int[]{5, 5, 5, 5}, 3));
    }

    @Test
    public void testNonUniformData() {
        // Skewed distribution — worst-case-like behavior
        assertEquals("Should still find 9 even in skewed data", 8, InterpolationSearch.interpolationSearch(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 1000}, 9));
    }

    @Test
    public void testDivisionByZeroProtection() {
        assertEquals("Equal elements — should handle safely", 0, InterpolationSearch.interpolationSearch(new int[]{7, 7, 7, 7}, 7));
        assertEquals("Equal elements — not found should return -1", -1, InterpolationSearch.interpolationSearch(new int[]{7, 7, 7, 7}, 5));
    }

    @Test
    public void testOutOfBoundsEstimation() {
        int result = InterpolationSearch.interpolationSearch(new int[]{10, 20, 30}, 25);
        assertTrue("Should not crash on bad estimate", result == -1 || result == 2); // Either not found or found via fallback
    }
}